http/client/src/main/java/org/wildfly/security/http/client/mechanism/digest/util/DigestHttpMechanismUtil.java

```diff
@@ -0,0 +1,155 @@
+        String cnonce = generateCNonce();
+        String nc = String.format("%08x", Integer.parseInt(authParams.getOrDefault("nc", String.valueOf(0))));
+
+        String resp;
+        if (qop == null) {
+            resp = computeDigestWithoutQop(uriPath, nonce, userName, password, algorithm, realm, "GET");
+        } else {
+            resp = computeDigestWithQop(uriPath, nonce, cnonce, nc, userName, password, algorithm, realm, qop, "GET");
+        }
+
+        StringBuilder requestAuthHeader = new StringBuilder();
+        requestAuthHeader.append("Digest ");
+        requestAuthHeader.append("username=\"").append(userName).append("\", ");
+        requestAuthHeader.append("realm=\"").append(realm).append("\", ");
+        requestAuthHeader.append("nonce=\"").append(nonce).append("\", ");
+        requestAuthHeader.append("uri=\"").append(uriPath).append("\", ");
+        requestAuthHeader.append("qop=\"").append(qop).append("\", ");
+        requestAuthHeader.append("nc=\"").append(nc).append("\", ");
+        requestAuthHeader.append("cnonce=\"").append(cnonce).append("\", ");
+        requestAuthHeader.append("response=\"").append(resp).append("\", ");
+        requestAuthHeader.append("opaque=\"").append(opaque).append("\", ");
+        requestAuthHeader.append("algorithm=").append(algorithm);
+
+        HttpRequest request = HttpRequest
+                .newBuilder()
+                .uri(uri)
+                .header(AUTHORIZATION, requestAuthHeader.toString())
+                .build();
+        int updateNonceCount = Integer.parseInt(authParams.getOrDefault("nc", "0")) + 1;
+        authParams.put("nc", String.valueOf(updateNonceCount));
+        return request;
+    }
+
+    private static Map<String,String> updateAuthParams(String authHeader) throws AuthenticationMechanismException {
+        byte[] rawHeader = authHeader.substring(CHALLENGE_PREFIX.length()).getBytes(UTF_8);
+        HashMap<String, byte[]> authval = parseResponse(rawHeader, UTF_8,false, log);
+        Map<String,String> authParams = new HashMap<>();
+        for(String headerKey : authval.keySet()){
+            String headerValue = new String(authval.get(headerKey),UTF_8);
+            authParams.put(headerKey,headerValue);
+        }
+        return authParams;
+    }
+
+    private static String computeDigestWithoutQop(String uri, String nonce, String username, String
+            password, String algorithm, String realm, String method) {
+        String A1, HashA1, A2, HashA2;
+        A1 = username + ":" + realm + ":" + password;
+        HashA1 = generateHash(A1,algorithm);
+        A2 = method + ":" + uri;
+        HashA2 = generateHash(A2,algorithm);
+        String combo, finalHash;
+        combo = HashA1 + ":" + nonce + ":" + HashA2;
+        finalHash = generateHash(combo,algorithm);
+        return finalHash;
+    }
+
+    private static String computeDigestWithQop(String uri, String nonce, String cnonce, String nc, String
```

Comment:
@keshav-725 Can DigestUtil class be used here as well? We already have sasl client computing this so these methods could be reused for http as well? 

Or can we refactor the classes so that both sasldigestclient and httpdigestclient use the same methods?

Reply:
Maybe moving the methods that both the SaslClient and HttpClient need to org.wildfly.security.mechanism.digest.DigestUtil makes sense?